packages/__tests__/3-runtime-html/controller.deactivation.partially-activated.spec.ts

```diff
@@ -0,0 +1,2677 @@
+        mgr.log.push(`${prefix}enter`);
+        return onResolve(
+          hook(initiator, parent),
+          () => { mgr.log.push(`${prefix}leave`); }
+        );
+      };
+    }
+
+    public createLifecycleHookInvoker(hookInstance: TestHook, hookName: string, hook: LifecycleHook): LifecycleHook {
+      const name = hookInstance.constructor.name;
+      hook = hook.bind(hookInstance);
+      const mgr = hookInstance.mgr;
+      return function $hook(vm: IRouteViewModel, initiator: IHydratedController, parent?: IHydratedController) {
+        const ctor = vm?.constructor;
+        const ceName = CustomElement.isType(ctor)
+          ? CustomElement.getDefinition(ctor).name
+          : CustomAttribute.isType(ctor)
+            ? CustomAttribute.getDefinition(ctor).name
+            : ctor?.name;
+        const prefix = `${mgr.prefix}.${name}.${ceName}.${hookName}.`;
+        mgr.log.push(`${prefix}enter`);
+        return onResolve(
+          hook(vm, initiator, parent),
+          () => { mgr.log.push(`${prefix}leave`); }
+        );
+      };
+    }
+
+    public assertLog(expected: string[], message: string) {
+      const log = this.log;
+      const len = Math.max(log.length, expected.length);
+      for (let i = 0; i < len; i++) {
+        assert.strictEqual(log[i], expected[i], `${message} - unexpected log at index${i}: ${log[i]}; actual log: ${JSON.stringify(log, undefined, 2)}`);
+      }
+    }
+
+    public setPrefix(prefix: string) {
+      this.prefix = prefix;
+      this.log.length = 0;
+    }
+  }
+
+  interface IPromiseManager {
+    createPromise(): Promise<void>;
+  }
+  const IPromiseManager = DI.createInterface<IPromiseManager>('IPromiseManager', x => x.singleton(PromiseManager));
+
+  type PromiseManagerMode = 'pending' | 'resolved' | 'rejected';
+  class PromiseManager implements IPromiseManager {
+    private _mode: PromiseManagerMode = 'pending';
+    private _resolve: () => void | null = null;
+    private _reject: (err: unknown) => void | null = null;
+    private _currentPromise: Promise<void> | null = null;
+    public get currentPromise(): Promise<void> | null { return this._currentPromise; }
+
+    public createPromise(): Promise<void> {
+      switch (this._mode) {
+        case 'pending': return this._currentPromise = new Promise<void>((res, rej) => { this._resolve = res; this._reject = rej; });
+        case 'resolved': return this._currentPromise = Promise.resolve();
+        case 'rejected': return this._currentPromise = Promise.reject();
+      }
+    }
+
+    public resolve() {
+      this._resolve?.();
+    }
+
+    public reject(err: unknown) {
+      this._reject?.(err);
+    }
+
+    public setMode(value: PromiseManagerMode): void {
+      this._mode = value;
+      this._resolve = null;
+      this._reject = null;
+    }
+  }
+
+  abstract class TestVM implements ICustomElementViewModel {
+    public constructor(
+      @INotifierManager public readonly mgr: INotifierManager,
+    ) {
```

Comment:
For simpler code, I think we can start using `resolve`
```suggestion
    public readonly mgr = resolve(INotifierManager);
    public constructor() {
```